src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -178,7 +178,9 @@ private[concurrent] object Promise {
    * DefaultPromises, and `linkedRootOf` is currently only designed to be called
    * by Future.flatMap.
    */
-  final class DefaultPromise[T] extends AtomicReference[AnyRef](Nil) with Promise[T] {
+  // Left non-final to enable addition of extra fields by Java/Scala converters
+  // in scala-java8-compat.
+  class DefaultPromise[T] extends AtomicReference[AnyRef](Nil) with Promise[T] {
```

Comment:
One thing to keep in mind is that this class extends AtomicReference (for performance reasons, since VarHandles are not available and AtomicReferenceUpdaters are slow) so subclasses could technically override the methods on AtomicReference (or worse, mess around with them).

I think we ought to put a big WARNING sign that says "don't mess around with the innards"